### PR TITLE
[SYCL][Graph] Add support for memory copy operations for CUDA backend

### PR DESCRIPTION
**Pull request by mfrancepillois**
Implements CUDA backend support for:
- memcpyUSM
- memcpy Device to Device (1D and 2D)
- memcpy Host to Device (Write 1D and 2D)
- memcpy Device to Host (Read 1D and 2D) 

Enables all remaining e2e tests

### CODE REVIEW

**Comment by EwanC** [comment]
> Enables all remaining e2e tests

I don't see this test begin enabled for CUDA https://github.com/reble/llvm/blob/080f8298a207d2efd91922f576718e2f28b821aa/sycl/test-e2e/Graph/device_query.cpp

Think we should test that native support is reported on CUDA now with this query.

**Comment by mfrancepillois** [comment]
> > Enables all remaining e2e tests
> 
> I don't see this test begin enabled for CUDA https://github.com/reble/llvm/blob/080f8298a207d2efd91922f576718e2f28b821aa/sycl/test-e2e/Graph/device_query.cpp
> 
> Think we should test that native support is reported on CUDA now with this query.

I pushed a new commit improving this test to verify that CUDA is reported as supported backend

File: sycl/plugins/unified_runtime/ur/adapters/cuda/command_buffer.cpp [hunk 1]
```diff
@@ -253,23 +313,33 @@ UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferAppendMembufferCopyRectExp(
...
+  ur_result_t Result;
+  CUgraphNode GraphNode;
+  std::vector<CUgraphNode> DepsList;
+  UR_CALL(getNodesFromSyncPoints(hCommandBuffer, numSyncPointsInWaitList,
+                                 pSyncPointWaitList, DepsList));
+
+  try {
+    CUdeviceptr SrcPtr = hSrcMem->Mem.BufferMem.get();
+    CUdeviceptr DstPtr = hDstMem->Mem.BufferMem.get();
+    CUDA_MEMCPY3D nodeParams = {};
```
Comment (EwanC) [inline sycl/plugins/unified_runtime/ur/adapters/cuda/command_buffer.cpp]: `NodeParams` PascalCase, and in other places

File: sycl/plugins/unified_runtime/ur/adapters/cuda/enqueue.hpp [hunk 2]
```diff
@@ -31,3 +31,11 @@ setKernelParams(const ur_context_handle_t Context,
...
                 CUfunction &CuFunc, size_t (&ThreadsPerBlock)[3],
                 size_t (&BlocksPerGrid)[3]);
+
+void setCopyRectParams(ur_rect_region_t region, const void *SrcPtr,
+                       const CUmemorytype_enum SrcType,
+                       ur_rect_offset_t src_offset, size_t src_row_pitch,
+                       size_t src_slice_pitch, void *DstPtr,
+                       const CUmemorytype_enum DstType,
+                       ur_rect_offset_t dst_offset, size_t dst_row_pitch,
+                       size_t dst_slice_pitch, CUDA_MEMCPY3D &params);
```
Comment (EwanC) [inline sycl/plugins/unified_runtime/ur/adapters/cuda/enqueue.hpp:41]: The variable naming convention here is inconsistent, but makes for the cleanest diff against upstream for the refactored code, which is nice when working in a fork. So agree to keep as-is, and if the maintainers would like us to change it when up streaming, we can address it then

**Review by EwanC — APPROVED** [review]
Looks great 😃

File: sycl/plugins/unified_runtime/ur/adapters/cuda/command_buffer.cpp [hunk 3]
```diff
@@ -230,19 +277,32 @@ UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferAppendMembufferCopyExp(
...
+
+    CUDA_MEMCPY3D nodeParams = {};
+    setCopyParams(&Src, CU_MEMORYTYPE_DEVICE, &Dst, CU_MEMORYTYPE_DEVICE, size,
+                  nodeParams);
+
+    Result = UR_CHECK_ERROR(cuGraphAddMemcpyNode(
+        &GraphNode, hCommandBuffer->CudaGraph, DepsList.data(), DepsList.size(),
+        &nodeParams, hCommandBuffer->Device->getContext()));
+
+    // Get sync point and register the event with it.
```
Comment (Bensuo) [inline sycl/plugins/unified_runtime/ur/adapters/cuda/command_buffer.cpp]: Here and in other places: `event` doesn't make sense in this context.

File: sycl/plugins/unified_runtime/ur/adapters/cuda/enqueue.cpp [hunk 4]
```diff
@@ -473,23 +473,16 @@ UR_APIEXPORT ur_result_t UR_APICALL urEnqueueKernelLaunch(
   return Result;
 }
 
-/// General 3D memory copy operation.
-/// This function requires the corresponding CUDA context to be at the top of
-/// the context stack
+/// Set parameter for General 3D memory copy.
```
Comment (Bensuo) [inline sycl/plugins/unified_runtime/ur/adapters/cuda/enqueue.cpp]:
```suggestion
/// Set parameters for general 3D memory copy.
```

**Review by Bensuo — APPROVED** [review]

File: sycl/plugins/unified_runtime/ur/adapters/cuda/command_buffer.cpp [hunk 5]
```diff
@@ -148,7 +183,7 @@ UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferAppendKernelLaunchExp(
           cuGraphAddEmptyNode(&GraphNode, hCommandBuffer->CudaGraph,
                               DepsList.data(), DepsList.size()));
 
-      // Get sync point and register the event with it.
+      // Get sync point and register the cuNode with it.
       *pSyncPoint = hCommandBuffer->GetNextSyncPoint();
       hCommandBuffer->RegisterSyncPoint(
           *pSyncPoint, std::make_shared<CUgraphNode>(GraphNode));
```
Comment (julianmi) [inline sycl/plugins/unified_runtime/ur/adapters/cuda/command_buffer.cpp]: Does it make sense to add a helper routine for the sync point to not have to duplicate this code throughout?

**Review by julianmi — APPROVED** [review]

File: sycl/plugins/unified_runtime/ur/adapters/cuda/command_buffer.cpp (hunk 5 above)
Comment (mfrancepillois) [reply sycl/plugins/unified_runtime/ur/adapters/cuda/command_buffer.cpp]: I added an helper to register sync points.

**Pull request merged**